File: Sources/App/Classes/Dialogs/Server Endpoint/TDCServerEndpointListSheet.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "IRCServer.h"
#import "TVCBasicTableView.h"
#import "TDCServerEndpointListSheetPrivate.h"

NS_ASSUME_NONNULL_BEGIN

static NSPasteboardType const _endpointEntryTableDragToken = @"com.vakesz.glasstual.server-endpoint-list.table-row";

@interface TDCServerEndpointListSheet ()
@property(nonatomic, strong) IBOutlet NSArrayController *entryTableController;
@property(nonatomic, weak) IBOutlet TVCBasicTableView *entryTable;
@property(nonatomic, weak) IBOutlet NSSegmentedControl *entryActionsSegmentedControl;

- (IBAction)entryActionsSegmentedControlClicked:(nullable id)sender;
@end

@implementation TDCServerEndpointListSheet

- (instancetype)initWithWindow:(nullable NSWindow *)window
{
	if ((self = [super initWithWindow:window])) {
		[self prepareInitialState];

		return self;
	}

	return nil;
}

- (void)prepareInitialState
{
	[RZMainBundle() loadNibNamed:@"TDCServerEndpointListSheet" owner:self topLevelObjects:nil];

	[self.entryTable registerForDraggedTypes:@[ _endpointEntryTableDragToken ]];
	self.entryTable.draggingDestinationFeedbackStyle = NSTableViewDraggingDestinationFeedbackStyleGap;

	[self.entryTableController addObserver:self
								forKeyPath:@"canRemove"
								   options:(NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew)
								   context:NULL];
}

- (void)startWithServerList:(NSArray<IRCServer *> *)serverList
{
	NSParameterAssert(serverList != nil);

	for (IRCServer *server in serverList) {
		[self.entryTableController addObject:[server mutableCopy]];
	}

	[self startSheet];
}

- (void)ok:(nullable id)sender
{
	NSArray *serverListIn = self.entryTableController.arrangedObjects;

	NSMutableArray<IRCServer *> *serverListOut = [[NSMutableArray alloc] initWithCapacity:serverListIn.count];

	for (IRCServerMutable *server in serverListIn) {
		/* New entries that are blank do not perform validation
		 because nothing technically has changed. Instead of 
		 doing some complex workaround, let's just ditch 
		 objects with an empty server address. */
		if (server.serverAddress.length == 0) {
			continue;
		}

		[serverListOut addObject:[server copy]];
	}

	if ([self.delegate respondsToSelector:@selector(serverEndpointListSheet:onOk:)]) {
		[self.delegate serverEndpointListSheet:self onOk:[serverListOut copy]];
	}

	[super ok:sender];
}

- (void)observeValueForKeyPath:(nullable NSString *)keyPath
					  ofObject:(nullable id)object
						change:(nullable NSDictionary<NSString *, id> *)change
					   context:(nullable void *)context
{
	if ([keyPath isEqualToString:@"canRemove"]) {
		[self updateEntryActionsSegmentedControlEnabledState];
	}
}

#pragma mark -
#pragma mark Entry Management

- (void)updateEntryActionsSegmentedControlEnabledState
{
	[self.entryActionsSegmentedControl setEnabled:self.entryTableController.canRemove forSegment:1];
}

- (void)entryActionsSegmentedControlClicked:(nullable id)sender
{
	NSInteger selectedSegment = [sender selectedSegment];

	if (selectedSegment == 0) {
		[self addEntry];
	} else if (selectedSegment == 1) {
		[self removeSelectedEntry];
	}
}

- (void)addEntry
{
	IRCServerMutable *newEntry = [IRCServerMutable new];

	[self.entryTableController addObject:newEntry];

	/* Edit column next pass on the main thread to allow the
	 -addObject to register properly. */
	XRPerformBlockAsynchronouslyOnMainQueue(^{
		NSTableView *tableView = self.entryTable;

		NSInteger rowSelection = (tableView.numberOfRows - 1);

		[tableView scrollRowToVisible:rowSelection];

		[tableView editColumn:0 row:rowSelection withEvent:nil select:YES];
	});
}

- (void)removeSelectedEntry
{
	NSIndexSet *selectedRows = self.entryTable.selectedRowIndexes;

	[self.entryTableController removeObjectsAtArrangedObjectIndexes:selectedRows];
}

#pragma mark -
#pragma mark Table View Delegate

- (nullable id<NSPasteboardWriting>)tableView:(NSTableView *)tableView pasteboardWriterForRow:(NSInteger)row
{
	NSPasteboardItem *item = [[NSPasteboardItem alloc] init];
	[item setString:@(row).stringValue forType:_endpointEntryTableDragToken];
	return item;
}

- (NSDragOperation)tableView:(NSTableView *)tableView
				validateDrop:(id<NSDraggingInfo>)info
				 proposedRow:(NSInteger)row
	   proposedDropOperation:(NSTableViewDropOperation)dropOperation
{
	/* Rows are reordered, never nested. Gap feedback requires the
	 drop to be expressed as above a row. */
	[tableView setDropRow:row dropOperation:NSTableViewDropAbove];

	return NSDragOperationMove;
}

- (BOOL)tableView:(NSTableView *)tableView
	   acceptDrop:(id<NSDraggingInfo>)info
			  row:(NSInteger)row
	dropOperation:(NSTableViewDropOperation)dropOperation
{
	NSPasteboard *pasteboard = [info draggingPasteboard];

	NSString *draggedRowString = [pasteboard stringForType:_endpointEntryTableDragToken];

	NSUInteger draggedRowIndex = draggedRowString.integerValue;

	[self.entryTableController moveObjectAtArrangedObjectIndex:draggedRowIndex toIndex:row];

	return YES;
}

#pragma mark -
#pragma mark NSWindow Delegate

- (void)windowWillClose:(NSNotification *)note
{
	[self.entryTableController removeObserver:self forKeyPath:@"canRemove"];

	if ([self.delegate respondsToSelector:@selector(serverEndpointListSheetWillClose:)]) {
		[self.delegate serverEndpointListSheetWillClose:self];
	}
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/AppKitSupport/TDCServerEndpointListSheet.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit
import CocoaExtensions

private let endpointEntryTableDragToken = NSPasteboard.PasteboardType(
	"com.vakesz.glasstual.server-endpoint-list.table-row"
)

/// What `ServerEndpointListSheet` reports back. The endpoints are values, so
/// they cannot travel through `NSObject.perform(_:with:with:)`.
@MainActor
public protocol ServerEndpointListSheetDelegate: AnyObject {
	func serverEndpointListSheet(_ sender: ServerEndpointListSheet, onOk serverList: [Server])
	func serverEndpointListSheetWillClose(_ sender: ServerEndpointListSheet)
}

@objc(TDCServerEndpointListSheet)
@MainActor
public final class ServerEndpointListSheet: SheetBase {
	@IBOutlet private var entryTableController: NSArrayController!
	@IBOutlet private var entryTable: BasicTableView!
	@IBOutlet private var entryActionsSegmentedControl: NSSegmentedControl!

	private var canRemoveObservation: NSKeyValueObservation?

	override public init(window: NSWindow?) {
		super.init(window: window)
		prepareInitialState()
	}

	private func prepareInitialState() {
		Bundle.main.loadNibNamed("TDCServerEndpointListSheet", owner: self, topLevelObjects: nil)

		entryTable.registerForDraggedTypes([endpointEntryTableDragToken])
		entryTable.draggingDestinationFeedbackStyle = .gap

		canRemoveObservation = entryTableController.observe(\.canRemove, options: [
			.initial,
			.new,
		]) { [weak self] _, _ in
			Task { @MainActor [weak self] in
				self?.updateEntryActionsSegmentedControlEnabledState()
			}
		}
	}

	private var endpointDelegate: (any ServerEndpointListSheetDelegate)? {
		delegate as? any ServerEndpointListSheetDelegate
	}

	public func start(with serverList: [Server]) {
		for server in serverList {
			entryTableController.addObject(ServerEndpointListEntry(server: server))
		}

		startSheet()
	}

	@IBAction override public func ok(_ sender: Any?) {
		let serverListIn = entryTableController.arrangedObjects as? [Any] ?? []
		let serverListOut = serverListIn
			.compactMap { ($0 as? ServerEndpointListEntry)?.server }
			.filter { $0.serverAddress.isEmpty == false }

		endpointDelegate?.serverEndpointListSheet(self, onOk: serverListOut)

		super.ok(sender)
	}

	private func updateEntryActionsSegmentedControlEnabledState() {
		entryActionsSegmentedControl.setEnabled(entryTableController.canRemove, forSegment: 1)
	}

	@IBAction private func entryActionsSegmentedControlClicked(_ sender: NSSegmentedControl) {
		switch sender.selectedSegment {
		case 0:
			addEntry()
		case 1:
			removeSelectedEntry()
		default:
			break
		}
	}

	private func addEntry() {
		entryTableController.addObject(ServerEndpointListEntry(server: Server()))

		DispatchQueue.main.async { [weak self] in
			guard let self else {
				return
			}

			let rowSelection = entryTable.numberOfRows - 1
			entryTable.scrollRowToVisible(rowSelection)
			entryTable.editColumn(0, row: rowSelection, with: nil, select: true)
		}
	}

	private func removeSelectedEntry() {
		let selectedRows = entryTable.selectedRowIndexes
		entryTableController.perform(NSSelectorFromString("removeObjectsAtArrangedObjectIndexes:"), with: selectedRows)
	}

	@objc public func windowWillClose(_: Notification) {
		canRemoveObservation?.invalidate()
		canRemoveObservation = nil

		endpointDelegate?.serverEndpointListSheetWillClose(self)
	}
}

extension ServerEndpointListSheet: NSTableViewDataSource, NSTableViewDelegate {
	@objc(tableView:pasteboardWriterForRow:)
	public func tableView(_: NSTableView, pasteboardWriterForRow row: Int) -> NSPasteboardWriting? {
		let item = NSPasteboardItem()
		item.setString(String(row), forType: endpointEntryTableDragToken)
		return item
	}

	@objc(
		tableView:validateDrop:proposedRow:proposedDropOperation:
	)
	public func tableView(
		_ tableView: NSTableView,
		validateDrop _: any NSDraggingInfo,
		proposedRow row: Int,
		proposedDropOperation _: NSTableView.DropOperation
	) -> NSDragOperation {
		tableView.setDropRow(row, dropOperation: .above)
		return .move
	}

	@objc(tableView:acceptDrop:row:dropOperation:)
	public func tableView(
		_: NSTableView,
		acceptDrop info: any NSDraggingInfo,
		row: Int,
		dropOperation _: NSTableView.DropOperation
	) -> Bool {
		guard
			let draggedRowString = info.draggingPasteboard.string(forType: endpointEntryTableDragToken),
			let draggedRowIndex = Int(draggedRowString)
		else {
			return false
		}

		entryTableController.textual_moveObject(atArrangedObjectIndex: UInt(draggedRowIndex), to: UInt(row))
		return true
	}
}
